### ISSUE TITLE
[Upcoming change]Enable Autocomplete filter in Edit Dataset setting by default

### ISSUE THREAD

**Issue by junlincc** [apache/superset#15694]
https://user-images.githubusercontent.com/67837651/125691871-f5d21b4f-fc30-4c4e-a0aa-fad37909000e.mov

We will be setting Autocomplete filter checkbox checked by default, as increasingly amount of new users report that 1) They don't know viewing filter values in the select dropdown is an option in Superset 2) navigating from chart to checkbox is unintuitive, as it's hidden in Edit dataset modal. 3) the purpose and functionality of the checkbox it's unclear 

Below changes will take place in steps
1.  By default, to enable autocomplete filter for better first time users experience - July 21st 
2. Adhoc filter popover redesign to view filter related settings in-place 2021 Q4
3. Edit dataset modal redesign to better organize different components and settings TBD

**Comment by etr2460** [comment]
A couple thoughts about the change here:

1. I believe the checkbox exists as the query run to generate the autocomplete can be very expensive on large tables, especially those with many partitions. It runs a query like:
```sql
SELECT DISTINCT "column_name" AS "column_name"
FROM "table_name"
 LIMIT 10000
```
which can be quite expensive, or return an error for our Presto setup (since you can filter on calculated columns/metrics which can contain arbitrarily complex sql queries). This is a bit less of a concern since you can still type in a value while the filter is running the query, but i'm not sure how the error state is handled here. We should verify that before changing the default. I'd also be concerned about removing the control entirely, since in some cases we probably want the ability to prevent superset from spamming the data warehouse with expensive, long running, synchronous queries .
2. A bigger concern I have is that the autocomplete only returns 10000 results and in a seemingly arbitrary order. This means that it can be difficult to select the actual value you want. See the gif below of what happens when you type in a value and press enter:
![Jul-15-2021 15-37-55](https://user-images.githubusercontent.com/7409244/125797481-328bc2f6-c2e0-40d6-beaf-f50ccf4e190d.gif)
Before making this the default, could we try to address some of these UX issues?
3. Finally, if we change the default here, I assume it's changing the default for newly created datasets. Would we want to migrate already existing datasets to default the checkbox to checked too?

Basically, I'm a bit worried about the increased query load this could add to our warehouse, error handling, and the UX of the current feature. While the happy path with small datasets seems to be fully functional, it would be good to address some of the large deployment edge cases before changing the default. Hope this helps!

**Comment by graceguo-supercat** [comment]
To make Superset works for both small and large dataset, could you add a config for this default behavior? Each company should have their own preference.

**Comment by junlincc** [comment]
understood and it sounds good. we will be add the config flag and update the UPDATING.md.

**Comment by rusackas** [comment]
Appreciate the detailed feedback/info on the concerns @etr2460. I also agree with @graceguo-supercat on the configurability. I also agree with @junlincc that it may make sense to start lumping some behaviors like this under a single flag. I'm not sure if we still have concerns about the "Open in Explore" link target, but I'll use it in these examples...

We could either be forward looking, or backwards compatible with the flag, e.g.:
* `BETA_UX_BEHAVIOR = True` could do things like opening Explore in the same window, and turning on autocomplete from this example
* `LEGACY_BEHAVIOR = False` could do exactly the inverse

Either way, it could be overridden, but maybe the Legacy branding makes more sense, as turning on a Beta flag by default sounds weird.

But would it make sense to folks to group this sort of thing under a single flag, or are we not worried about a potential proliferation of flags for single UX/UI items like these?

**Comment by michael-s-molina** [comment]
> 1. I believe the checkbox exists as the query run to generate the autocomplete can be very expensive on large tables, especially those with many partitions.

> 2. A bigger concern I have is that the autocomplete only returns 10000 results and in a seemingly arbitrary order. This means that it can be difficult to select the actual value you want. 

@etr2460 Both points are very important. One thing that we can do to mitigate these risks is to use the new Select component. It will allow us to fetch data in a paginated manner and also remove the limit on the number of results. We can submit a PR and test in Airbnb's environment to check if performance is acceptable 😉.

**Comment by graceguo-supercat** [comment]
have a `blanket flag`, like LEGACY_BEHAVIOR, i think it could work. Please let us know the list of items that you want to change its default.

**Comment by etr2460** [comment]
>One thing that we can do to mitigate these risks is to use the new Select component. It will allow us to fetch data in a paginated manner and also remove the limit on the number of results. We can submit a PR and test in Airbnb's environment to check if performance is acceptable 😉.

@michael-s-molina Pagination could help in some cases, but not in all of them. This would be especially difficult for virtual datasets, since the limit for pagination would come after the full virtual dataset query completes in Presto (for example if there's an order by in the virtual dataset query). Honestly, this is something really hard, and possibly would be better as a cached/pre-warmed set of metadata on the dataset itself. Something like `common_column_values` in the metric/column metadata that is warmed on a cadence set by the user/admin. idk, just spitballing here, but this feature in general becomes a larger problem the larger a deployment gets. It probably doesn't scale to Airbnb's deployment in the current state (which is why it's defaulted to off :P )

**Comment by michael-s-molina** [comment]
> @michael-s-molina Pagination could help in some cases, but not in all of them. This would be especially difficult for virtual datasets, since the limit for pagination would come after the full virtual dataset query completes in Presto (for example if there's an order by in the virtual dataset query). Honestly, this is something really hard, and possibly would be better as a cached/pre-warmed set of metadata on the dataset itself. Something like `common_column_values` in the metric/column metadata that is warmed on a cadence set by the user/admin. idk, just spitballing here, but this feature in general becomes a larger problem the larger a deployment gets. It probably doesn't scale to Airbnb's deployment in the current state (which is why it's defaulted to off :P )

@etr2460 Thank you for the additional context. I agree with you that for that type of deployment we'll need to have a configurable way to preserve the current behavior of only requesting the data after a search has been entered. For other types of deployments, we can benefit from pagination and lazy loading. Make this feature configurable is a win/win.

**Comment by ktmud** [comment]
> Appreciate the detailed feedback/info on the concerns @etr2460. I also agree with @graceguo-supercat on the configurability. I also agree with @junlincc that it may make sense to start lumping some behaviors like this under a single flag. I'm not sure if we still have concerns about the "Open in Explore" link target, but I'll use it in these examples...
> 
> We could either be forward looking, or backwards compatible with the flag, e.g.:
> 
> * `BETA_UX_BEHAVIOR = True` could do things like opening Explore in the same window, and turning on autocomplete from this example
> * `LEGACY_BEHAVIOR = False` could do exactly the inverse
> 
> Either way, it could be overridden, but maybe the Legacy branding makes more sense, as turning on a Beta flag by default sounds weird.
> 
> But would it make sense to folks to group this sort of thing under a single flag, or are we not worried about a potential proliferation of flags for single UX/UI items like these?

+1 on grouping new UX behaviors under a single feature flags. But may I suggest calling them `EXPERIMENTAL_UX_BEHAVIORS` and  keep them off by default? We can keep add new features or changing default behaviors _without_ this flag and only move them into the flag when it becomes a contentious point. Most of the time, when people don't want a feature, it is because there are some shortcomings in the implementation---even though they may benefit the majority of Superset users. For example, the open chart in same window would be nice to have even for Airbnb if cmd + click is possible AND navigating back to dashboard does not reload the whole page; select options default on is good for everyone if it won't cause db stress for large datasets & it can show all options... I think it's fair to keep them behind a "EXPERIMENTAL" flag until we address these addressable shortcomings.

**Comment by junlincc** [comment]
@kgabryje let's add this change to UX_BETA, along with preselected aggregation.
